### PR TITLE
Caching of Pointers to Redzone Checker Kernel

### PR DESCRIPTION
**Pull request by philipphack**
Introduces a cache for pointers to the Redzone Checker kernel which prevents the repeated loading and unloading of the kernel.

### CODE REVIEW

**Comment by timshen91** [comment]
Unassign myself as I don't have time and I'm already reviewing other patches. @gbaned can you reassign?

**Comment by gbaned** [comment]
@philipphack  Can you please resolve conflicts? Thanks!

File: tensorflow/stream_executor/gpu/asm_compiler.h [hunk 1]
```diff
@@ -18,10 +18,15 @@ limitations under the License.
...
 
+#include "absl/container/flat_hash_map.h"
 #include "absl/types/span.h"
+#include "tensorflow/core/platform/mutex.h"
+#include "tensorflow/stream_executor/cuda/cuda_driver.h"
 #include "tensorflow/stream_executor/gpu/gpu_asm_opts.h"
+#include "tensorflow/stream_executor/kernel.h"
 #include "tensorflow/stream_executor/lib/statusor.h"
 #include "tensorflow/stream_executor/platform/port.h"
+#include "tensorflow/stream_executor/stream_executor_pimpl.h"
```
Comment (akuegel) [inline tensorflow/stream_executor/gpu/asm_compiler.h:29]: It looks to me that you might need to add a BUILD file dependency for this header include.
Probably "//tensorflow/stream_executor:kernel"

**Comment by akuegel** [comment]
@cheshire I think you are more familiar with this code, so I have added you as a reviewer, too.

**Review by cheshire — APPROVED** [review]

**Comment by cheshire** [comment]
This breaks ROCm build:

```
./tensorflow/stream_executor/gpu/asm_compiler.h:24:10: fatal error: tensorflow/stream_executor/cuda/cuda_driver.h: No such file or directory
 #include "tensorflow/stream_executor/cuda/cuda_driver.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Note that asm_compiler is used for both ROCm and CUDA.

**Comment by gbaned** [comment]
@philipphack  Can you please check @cheshire's comments and keep us posted ? Thanks!

**Review by cheshire — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

**Pull request merged**